internal/build/tar.go

```diff
@@ -232,6 +233,30 @@ func (a *ArchiveBuilder) writeEntry(entry archiveEntry) error {
 		_ = file.Close()
 	}()
 
+	// The size header must match the number of contents bytes.
+	//
+	// There is room for a race condition here if something writes to the file
+	// after we've read the file size.
+	//
+	// For small files, we avoid this by first copying the file into a buffer,
+	// and using the size of the buffer to populate the header.
+	//
+	// For larger files, we don't want to copy the whole thing into a buffer,
+	// because that would blow up heap size. There is some danger that this
+	// will lead to a spurious error when the tar writer validates the sizes.
+	// That error will be disruptive but will be handled as best as we
+	// can downstream.
+	useBuf := header.Size < 5000000
+	var buf *bytes.Buffer
+	if useBuf {
+		buf = bytes.NewBuffer(make([]byte, 0, header.Size))
```

Comment:
Arguably a premature optimization, but given that this will likely be the predominate case in most projects, worth storing the buf on `ArchiveBuilder` and re-using it?

Reply:
good point! Also added a test case to catch this issue